lib/std/os/windows.zig

```diff
@@ -2221,7 +2236,24 @@ pub fn wToPrefixedFileW(dir: ?HANDLE, path: [:0]const u16) !PathSpace {
                 // canonicalize it. We do this by getting the path of the `dir`
                 // and appending the relative path to it.
                 var dir_path_buf: [PATH_MAX_WIDE:0]u16 = undefined;
-                const dir_path = try GetFinalPathNameByHandle(dir.?, .{}, &dir_path_buf);
+                const dir_path = GetFinalPathNameByHandle(dir.?, .{}, &dir_path_buf) catch |err| switch (err) {
+                    // This mapping is not correct; it is actually expected
+                    // that calling GetFinalPathNameByHandle might return
+                    // error.UnrecognizedVolume, and in fact has been observed
+                    // in the wild. The problem is that wToPrefixedFileW was
+                    // never intended to make *any* OS syscall APIs. It's only
+                    // supposed to convert a string to one that is eligible to
+                    // be used in the ntdll syscalls.
+                    //
+                    // To solve this, this function needs to no longer call
+                    // GetFinalPathNameByHandle under any conditions, or the
+                    // calling function needs to get reworked to not need to
+                    // call this function.
+                    //
+                    // This may involve making breaking API changes.
```

Comment:
For context, the dilemma is:
- ntdll syscalls cannot handle paths with `..` components
- In the best case, we can remove `..` components trivially (this is what `normalizePath`/`removeDotDirsSanitized` does)
- However, to be able to make a syscall when given a path like `../../foo`, it needs to be resolved into an absolute path and given the NT namespace prefix (`\??\`)
- `RtlGetFullPathName_U` can handle the case when resolving a path like `../../foo` against the cwd, but not against an arbitrary directory handle
- Therefore, we use our own `GetFinalPathNameByHandle` implementation in the case where we (1) can't trivially remove `..` components and (2) need to resolve the path against a non-CWD dir

Reply:
> However, to be able to make a syscall when given a path like ../../foo, it needs to be resolved into an absolute path and given the NT namespace prefix (\??\)

Zig file system API should make this fail for all operating systems. You shouldn't be able to open a parent directory from an open directory handle.

The only exception is the current working directory, which is already an exception for other use cases.

This hack should have never been made, and by not making the hack it would have helped us reach that conclusion earlier.

Reply:
Fwiw, I believe it was me who reported ".. doesn’t work on windows” (https://github.com/ziglang/zig/issues/16779), and, for my use-case, it is indeed enough if that only works for cwd

Reply:
> Zig file system API should make this fail for all operating systems. You shouldn't be able to open a parent directory from an open directory handle.

This might be a bit complicated due to things like https://github.com/ziglang/zig/issues/18452 and https://github.com/ziglang/zig/issues/7751. That is, resolving `..` before symlinks (the Windows way) would be a break from normal non-Windows behavior, and resolving symlinks before `..` would require syscalls. Symlinks in general complicate things, too.

There is a potentially relevant `RESOLVE_BENEATH` flag on Linux (and `O_RESOLVE_BENEATH` on FreeBSD) but it is even more restrictive (does not support absolute paths): https://man7.org/linux/man-pages/man2/openat2.2.html